### DISCUSSION TITLE
Reliability of direct search (server bugs, or wrong expectations ?)

### DISCUSSION THREAD

**Discussion by hwti** [Tensai75/nzb-monkey-go#29] [General]
I have weird issues, like the search being randomly stuck, or errors (with an older version of nzb-monkey-go, it seems better now).
This lead me to try to investigate by capturing the exchanges with Wireshark, and trying manual commands with telnet.

I don't know if what I saw is specific to my provider, and what to expect from the NNTP servers in general.

Sometimes the server doesn't respond to a XOVER command, and everything is stuck.
I even managed to reproduce it by sending the same command several times with telnet, it seems really random, but in all cases sending the command again works.
I don't know if introducing a delay between commands would help, but it would slow down everything.
Retrying after a timeout would probably work, but it would be weird on a connection which is supposed to only get one command at a time (but i has to be on the same connection, see the second point below).

It seems there is some kind of load-balancing between non-synchronized servers.
So two connections to the same IP could lead to slightly different ranges returned by the GROUP command, and the same article number corresponding to different articles (ie different responses for the same XOVER command).
I don't know what use case would trigger the use of several connections, but unless all servers have common article numbers (Are there providers for which this is true ?), I think only one connection should be used for the whole search.

**Comment by Tensai75** [comment]
> I have weird issues, like the search being randomly stuck, or errors
> Sometimes the server doesn't respond to a XOVER command, and everything is stuck. 

What error messages (from the NZB Monkey Go) do you get in these cases?
And when exactly do the errors occur? During the search for the first or last message or during the actual scan?

> I don't know if what I saw is specific to my provider, and what to expect from the NNTP servers in general.

What provider are you using?

> So two connections to the same IP could lead to slightly different ranges returned by the GROUP command, 

This is perfectly normal, because between two consecutive GROUP commands, old articles may have been deleted and new articles added to the group. Therefore you can never expect the low and high value of the response from the GROUP command to be identical between two calls.

> and the same article number corresponding to different articles (ie different responses for the same XOVER command).
> I don't know what use case would trigger the use of several connections, but unless all servers have common article numbers (Are there providers for which this is true ?), I think only one connection should be used for the whole search.

No, the article _numbers_ of the articles within a group remain always the same. So, a `XOVER 1000-1999` command will always result in the exact same articles every time the command is executed, if this range is within the available articles range (within the low and high boundary).
So it makes very much sense to use several connections to download several articles in parallel with the XOVER command using different ranges (e.g. 1000-1999, 2000-2999, etc.).

**Reply by hwti** [reply]
> What error messages (from the NZB Monkey Go) do you get in these cases? And when exactly do the errors occur? During the search for the first or last message or during the actual scan?

I only got the errors with an older version, with the servers replying empty responses.
Maybe something has changed server-side, or now the requests are slightly different and do not reproduce the issue.

With the current version it can get stuck at any point :
 - during the search for first / last messages : I know this is the server not responding to the XOVER command
 - during the scan : I'm not sure what happened

> What provider are you using?

Frugal Usenet (NetNews US, the EU servers always give empty responses for XOVER or XHDR commands)

> This is perfectly normal, because between two consecutive GROUP commands, old articles may have been deleted and new articles added to the group. Therefore you can never expect the low and high value of the response from the GROUP command to be identical between two calls.

I saw a case where the low value decreased, as if each connection could be directed to a different physical server, and that the different servers wouldn't be updated at the same time.

> No, the article _numbers_ of the articles within a group remain always the same. So, a `XOVER 1000-1999` command will always result in the exact same articles every time the command is executed, if this range is within the available articles range (within the low and high boundary). So it makes very much sense to use several connections to download several articles in parallel with the XOVER command using different ranges (e.g. 1000-1999, 2000-2999, etc.).

Well, maybe I didn't pay attention to the number in the response, and was mislead by the behavior when getting ranges close from the lower limit.
```
GROUP alt.binaries.boneless
211 46339440004 42347403655 88686843658 alt.binaries.boneless
XOVER 42347403655-42347404655
224 xover information follows
42347403965
...
42347404655
```
```
GROUP alt.binaries.boneless
211 46339440004 42347403655 88686843658 alt.binaries.boneless
XOVER 42347403655-42347404655
224 xover information follows
42347404310
...
42347404655
```
Here the GROUP response was the same for the two connections, but not the XOVER response.
Maybe the first messages (42347403965-42347404309) were removed between the GROUP and XOVER command from the second connection, or maybe the response for the GROUP command was out of date.

```
GROUP alt.binaries.boneless
211 46339263570 42347401873 88686665442 alt.binaries.boneless
XOVER 42347401873-42347401883
224 xover information follows
42347402045 ...
```
Here I requested a small range starting from the low number, and I got a reply with just one message, but out of range (maybe the first message the server has).
```
GROUP alt.binaries.boneless
211 46339263570 42347401873 88686665442 alt.binaries.boneless
XOVER 42347401873-42347401883
224 xover information follows
42347402433 ...
```
The second connection reports the same low number for the GROUP command (if 42347402045 was actually the first message the server had, the GROUP command should report it or even an higher number, but it's still 42347401873).
And the replied message is newer.

```
GROUP alt.binaries.boneless
211 46339679965 42347406079 88687086043 alt.binaries.boneless
XOVER 42347403655-42347404655
224 xover information follows
42347406074 ...
```
This time I just copied a command from a previous test, so the XOVER argument if out of the range returned by GROUP.
But the server replied with 42347406074 (again, not in the range I asked), which is older than what it just claimed to have (42347406079) !

**Reply by Tensai75** [reply]
Unfortunately, I don't have a Frugal account to test this myself, but this behaviour seems very strange to me. Especially when the XOVER requests results in an article outside the requested range. I don't recall seeing this in my own testing (mostly on Newshosting/Eweka), but in any case, I don't think this is RFC 3977 compliant.

Anyway, I've already found out earlier that all commands that _**don't**_ directly request the actual body/full article, e.g. (X)OVER, HEAD, STAT and apparently GROUP, are sometimes quite unreliable. It seems that the results are pulled from the internal index of the news servers, which obviously can be out of sync with the actual availability of the articles. So it is possible (and I have experienced it myself) that the NZB Monkey Go can generate a "complete" NZB file from the (X)OVER overviews, although some or all articles are actually not available on the server at all. And as a result, the download with this NZB file will fail due to missing articles.

Unfortunately, I don't see any feasible way to overcome this unreliability...

> With the current version it can get stuck at any point :
> 
> * during the search for first / last messages : I know this is the server not responding to the XOVER command
> * during the scan : I'm not sure what happened

What exyctly do you mean by "get stuck"? Does the program simply freeze? Did you activate the debug log and if yes, is there something in the log file?

**Reply by hwti** [reply]
> Anyway, I've already found out earlier that all commands that _**don't**_ directly request the actual body/full article, e.g. (X)OVER, HEAD, STAT and apparently GROUP, are sometimes quite unreliable. It seems that the results are pulled from the internal index of the news servers, which obviously can be out of sync with the actual availability of the articles. So it is possible (and I have experienced it myself) that the NZB Monkey Go can generate a "complete" NZB file from the (X)OVER overviews, although some or all articles are actually not available on the server at all. And as a result, the download with this NZB file will fail due to missing articles.

There is also XHDR, but I suppose this is going to be similar.
It could still be an optimization (XHR DATE when searching for first / last message, and maybe XHDR SUBJECT for the scan), but only to reduce the size of what is downloaded.

> What exyctly do you mean by "get stuck"? Does the program simply freeze? Did you activate the debug log and if yes, is there something in the log file?

The log file doesn't give more information than the console (maybe I did something wrong).

The program freezes (the progress stops, ...) :
 - during the search for first / last message, nothing seems to happen
 - during the scan, there is still one thread fully using one core

It is waiting for the server to respond to a command which seems to have been "lost".

For example, a freeze on the first OVER command, then 5 min server timeout :
```
   Scanning for first message number ... 100% |████████████████████████████████████████|  [5m2s]
   ERROR:    Error while scanning group 'alt.binaries.multimedia.aviation' for the first message: EOF
```
From Wireshark :
```
GROUP alt.binaries.multimedia.aviation
211 1572865751 19064256 1591930006 alt.binaries.multimedia.aviation
OVER 805497131-805498131
(a TCP ACK from the server, then TCP keep-alives for 5 min)
400 Idle timeout.
XOVER 805497131-805498131
QUIT
```
The server behaves as if the command was never sent, even if the TCP ACK tells us that the packet has been received.
I have seen this behavior with OVER (not supported, the server usually returns an error) and XOVER.
When I reproduced similar cases manually with telnet, sending the command again worked.

**Comment by hwti** [comment]
Btw, both Windows and Linux versions saturate one core when scanning messages (with one connection), so I thought this was CPU limited.
But the Linux version seems to be quite faster, so maybe the CPU isn't doing actual work and there is a issue (polling ?).
